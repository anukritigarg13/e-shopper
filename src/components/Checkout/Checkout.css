input {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 50%;
    padding: 12px 20px;
    margin: 8px 0;
    box-sizing: border-box;
  }