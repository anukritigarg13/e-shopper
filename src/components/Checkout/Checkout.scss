
.checkout{
  display: flex;
  flex-direction: column;
  align-items: center;
}
.form{
  input{
    width: 100%;
    padding: 12px 20px;
    margin: 8px 0;
    box-sizing: border-box;
  }
  margin :5%
}
.checkout-message{
  margin: 4rem;
}