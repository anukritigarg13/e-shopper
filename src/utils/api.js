/* eslint-disable import/prefer-default-export */
import axios from 'axios';
import groupByCategory from './helper';

export const getAllItems = async (endpoint) => {
  const { data, error } = await axios.get(endpoint);
  if (data) {
    return groupByCategory(data.data);
  }
  throw new Error(error.message);
};

export const getAllOrders = async (endpoint) => {
  const { data: orders, error } = await axios.get(endpoint);
  if (orders) {
    const categorisedOrders = orders.data.reduce((allOrders, order) => {
      const categorisedOrderItems = groupByCategory(order.items);
      const totalCost = order.items
        .reduce((acc, product) => acc + product.count * product.price, 0);
      const newOrder = {
        ...order, items: categorisedOrderItems, totalItems: order.items.length, totalCost,
      };
      return allOrders.concat(newOrder);
    }, []);
    return categorisedOrders;
  }
  throw new Error(error.message);
};

export const postOrder = async (endpoint, products) => {
  const orderedProducts = Object.keys(products).reduce((accumulator, productCategory) => {
    const productsInCart = products[productCategory]
      .filter((product) => product.itemCount > 0).reduce((acc, product) => {
        const newCount = product.itemCount;
        const newProduct = { ...product, count: newCount };
        delete newProduct.itemCount;
        return acc.concat(newProduct);
      }, []);
    if (accumulator.items === undefined) {
      accumulator.items = [];
    }
    accumulator.items = accumulator.items.concat(productsInCart);
    return accumulator;
  }, {});
  const response = await axios.post('/orders', orderedProducts);
  return response;
};
