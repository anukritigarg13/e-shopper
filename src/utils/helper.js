const groupByCategory = (items) => {
  const categorisedItems = items.reduce((acc, product) => {
    if (acc[product.category] === undefined) {
      acc[product.category] = [];
    }
    const newProduct = { ...product };
    newProduct.itemCount = 0;
    acc[product.category].push(newProduct);
    return acc;
  }, {});
  return categorisedItems;
};

export default groupByCategory;
