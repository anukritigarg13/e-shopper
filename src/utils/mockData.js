/* eslint-disable import/prefer-default-export */
export const mockOrders = {
  data: [
    {
      items: [
        {
          id: 1,
          name: 'apple',
          price: 120,
          count: 1,
          category: 'Fruits & Vegatables',
        },
      ],
      id: 1,
      date: 1615122360481,
    },
    {
      items: [
        {
          id: 2,
          name: 'table cloth',
          price: 200,
          count: 1,
          category: 'Household Items',
        },
        {
          id: 9,
          name: 'broom',
          price: 250,
          count: 1,
          category: 'Household Items',
        },
      ],
      id: 2,
      date: 1615122649963,
    },
    {
      items: [
        {
          id: 1,
          name: 'apple',
          price: 120,
          count: 1,
          category: 'Fruits & Vegatables',
        },
        {
          id: 2,
          name: 'table cloth',
          price: 200,
          count: 1,
          category: 'Household Items',
        },
        {
          id: 9,
          name: 'broom',
          price: 250,
          count: 1,
          category: 'Household Items',
        },
        {
          id: 8,
          name: 'curd',
          price: 20,
          count: 2,
          category: 'Dairy & Eggs',
        },
      ],
      id: 3,
      date: 1615122664245,
    },
    {
      items: [
        {
          id: 4,
          name: 'duster',
          price: 80,
          count: 1,
          category: 'Household Items',
        },
        {
          id: 5,
          name: 'milk',
          price: 10,
          count: 2,
          category: 'Dairy & Eggs',
        },
        {
          id: 6,
          name: 'butter',
          price: 20,
          count: 2,
          category: 'Dairy & Eggs',
        },
      ],
      id: 4,
      date: 1615122763596,
    },
    {
      items: [
        {
          id: 1,
          name: 'apple',
          price: 120,
          count: 3,
          category: 'Fruits & Vegatables',
        },
        {
          id: 3,
          name: 'grapes',
          price: 50,
          count: 3,
          category: 'Fruits & Vegatables',
        },
        {
          id: 7,
          name: 'banana',
          price: 5,
          count: 2,
          category: 'Fruits & Vegatables',
        },
      ],
      id: 5,
      date: 1615540977063,
    },
  ],
};

export const mockProducts = {
  data: [
    {
      id: 1,
      name: 'apple',
      price: 120,
      count: 20,
      category: 'Fruits & Vegatables',
    },
    {
      id: 2,
      name: 'table cloth',
      price: 200,
      count: 3,
      category: 'Household Items',
    },
    {
      id: 3,
      name: 'grapes',
      price: 50,
      count: 10,
      category: 'Fruits & Vegatables',
    },
    {
      id: 4,
      name: 'duster',
      price: 80,
      count: 5,
      category: 'Household Items',
    },
    {
      id: 5,
      name: 'milk',
      price: 10,
      count: 2,
      category: 'Dairy & Eggs',
    },
    {
      id: 6,
      name: 'butter',
      price: 20,
      count: 2,
      category: 'Dairy & Eggs',
    },
    {
      id: 7,
      name: 'banana',
      price: 5,
      count: 48,
      category: 'Fruits & Vegatables',
    },
    {
      id: 8,
      name: 'curd',
      price: 20,
      count: 4,
      category: 'Dairy & Eggs',
    },
    {
      id: 9,
      name: 'broom',
      price: 250,
      count: 4,
      category: 'Household Items',
    },
    {
      id: 10,
      name: 'soap',
      price: 10,
      count: 0,
      category: 'Household Items',
    },
  ],
};

export const categorisedProducts = {
  'Fruits & Vegatables': [
    {
      id: 1,
      name: 'apple',
      price: 120,
      count: 20,
      category: 'Fruits & Vegatables',
      itemCount: 0,
    },
    {
      id: 3,
      name: 'grapes',
      price: 50,
      count: 10,
      category: 'Fruits & Vegatables',
      itemCount: 0,
    },
    {
      id: 7,
      name: 'banana',
      price: 5,
      count: 48,
      category: 'Fruits & Vegatables',
      itemCount: 0,
    },
  ],
  'Household Items': [
    {
      id: 2,
      name: 'table cloth',
      price: 200,
      count: 3,
      category: 'Household Items',
      itemCount: 0,
    },
    {
      id: 4,
      name: 'duster',
      price: 80,
      count: 5,
      category: 'Household Items',
      itemCount: 0,
    },
    {
      id: 9,
      name: 'broom',
      price: 250,
      count: 4,
      category: 'Household Items',
      itemCount: 0,
    },
    {
      id: 10,
      name: 'soap',
      price: 10,
      count: 0,
      category: 'Household Items',
      itemCount: 0,
    },
  ],
  'Dairy & Eggs': [
    {
      id: 5,
      name: 'milk',
      price: 10,
      count: 2,
      category: 'Dairy & Eggs',
      itemCount: 0,
    },
    {
      id: 6,
      name: 'butter',
      price: 20,
      count: 2,
      category: 'Dairy & Eggs',
      itemCount: 0,
    },
    {
      id: 8,
      name: 'curd',
      price: 20,
      count: 4,
      category: 'Dairy & Eggs',
      itemCount: 0,
    },
  ],
};

export const categorisedOrders = [
  {
    items: {
      'Fruits & Vegatables': [
        {
          id: 1,
          name: 'apple',
          price: 120,
          count: 1,
          category: 'Fruits & Vegatables',
          itemCount: 0,
        },
      ],
    },
    id: 1,
    date: 1615122360481,
    totalItems: 1,
    totalCost: 120,
  },
  {
    items: {
      'Household Items': [
        {
          id: 2,
          name: 'table cloth',
          price: 200,
          count: 1,
          category: 'Household Items',
          itemCount: 0,
        },
        {
          id: 9,
          name: 'broom',
          price: 250,
          count: 1,
          category: 'Household Items',
          itemCount: 0,
        },
      ],
    },
    id: 2,
    date: 1615122649963,
    totalItems: 2,
    totalCost: 450,
  },
  {
    items: {
      'Fruits & Vegatables': [
        {
          id: 1,
          name: 'apple',
          price: 120,
          count: 1,
          category: 'Fruits & Vegatables',
          itemCount: 0,
        },
      ],
      'Household Items': [
        {
          id: 2,
          name: 'table cloth',
          price: 200,
          count: 1,
          category: 'Household Items',
          itemCount: 0,
        },
        {
          id: 9,
          name: 'broom',
          price: 250,
          count: 1,
          category: 'Household Items',
          itemCount: 0,
        },
      ],
      'Dairy & Eggs': [
        {
          id: 8,
          name: 'curd',
          price: 20,
          count: 2,
          category: 'Dairy & Eggs',
          itemCount: 0,
        },
      ],
    },
    id: 3,
    date: 1615122664245,
    totalItems: 4,
    totalCost: 610,
  },
  {
    items: {
      'Household Items': [
        {
          id: 4,
          name: 'duster',
          price: 80,
          count: 1,
          category: 'Household Items',
          itemCount: 0,
        },
      ],
      'Dairy & Eggs': [
        {
          id: 5,
          name: 'milk',
          price: 10,
          count: 2,
          category: 'Dairy & Eggs',
          itemCount: 0,
        },
        {
          id: 6,
          name: 'butter',
          price: 20,
          count: 2,
          category: 'Dairy & Eggs',
          itemCount: 0,
        },
      ],
    },
    id: 4,
    date: 1615122763596,
    totalItems: 3,
    totalCost: 140,
  },
  {
    items: {
      'Fruits & Vegatables': [
        {
          id: 1,
          name: 'apple',
          price: 120,
          count: 3,
          category: 'Fruits & Vegatables',
          itemCount: 0,
        },
        {
          id: 3,
          name: 'grapes',
          price: 50,
          count: 3,
          category: 'Fruits & Vegatables',
          itemCount: 0,
        },
        {
          id: 7,
          name: 'banana',
          price: 5,
          count: 2,
          category: 'Fruits & Vegatables',
          itemCount: 0,
        },
      ],
    },
    id: 5,
    date: 1615540977063,
    totalItems: 3,
    totalCost: 520,
  },
];
