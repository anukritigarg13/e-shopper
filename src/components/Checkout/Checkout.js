import React, { useState } from 'react';
import './Checkout.scss';
import validator from 'validator';
import PropTypes from 'prop-types';
import { postOrder } from '../../utils/api';

const Checkout = ({ products, updateAllOrders, cartItemsCount }) => {
  const [name, setName] = useState('');
  const [contactNo, setContactNo] = useState(undefined);
  const [email, setEmail] = useState('');
  const [address, setAddress] = useState('');
  const [errorMessage, setErrorMessage] = useState(undefined);
  const [successMessage, setSuccessMessage] = useState(undefined);
  const onChangeHandler = (event, field) => {
    const newValue = event.target.value;
    if (field === 'contactNo' && newValue.length > 10) {
      return;
    }
    switch (field) {
      case ('name'):
        setName(newValue);
        break;
      case ('contactNo'):
        setContactNo(newValue);
        break;
      case ('address'):
        setAddress(newValue);
        break;
      case ('email'):
        setEmail(newValue);
        break;
      default:
    }
  };

  const onCheckoutHandler = async () => {
    if (!validator.isEmail(email)
     || name.length === 0 || address.length === 0 || !contactNo || contactNo.length < 10) {
      setErrorMessage('Please add valid details');
      setSuccessMessage(undefined);
      return;
    }
    if (cartItemsCount === 0) {
      setErrorMessage('Please add items to your cart');
      setSuccessMessage(undefined);
      return;
    }
    const response = await postOrder('/orders', products);
    updateAllOrders(response.data);
    setSuccessMessage('Thankyou for shopping with us!');
    setErrorMessage(undefined);
  };
  let message = '';
  if (errorMessage) message = errorMessage;
  else if (successMessage) message = successMessage;
  return (

    <div className="checkout">
      <div className="form">
        <form data-testid="form">
          Name
          <input type="text" onChange={(event) => onChangeHandler(event, 'name')} value={name} />
          Address
          <input type="text" onChange={(event) => onChangeHandler(event, 'address')} value={address} />
          Contact Number
          <input type="number" onChange={(event) => onChangeHandler(event, 'contactNo')} value={contactNo} />
          E-mail
          <input type="text" onChange={(event) => onChangeHandler(event, 'email')} value={email} />
        </form>
      </div>
      <div>
        <button onClick={onCheckoutHandler} type="button">Submit</button>
      </div>
      <div className="checkout-message">
        <h2>{message}</h2>
      </div>

      <div />

    </div>

  );
};
const productsShape = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  count: PropTypes.number.isRequired,
  category: PropTypes.string.isRequired,
});
Checkout.propTypes = {
  products: PropTypes.objectOf(PropTypes.arrayOf(productsShape)).isRequired,
  updateAllOrders: PropTypes.func.isRequired,
  cartItemsCount: PropTypes.number.isRequired,
};
export default Checkout;
