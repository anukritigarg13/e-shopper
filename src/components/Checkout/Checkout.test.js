/* eslint-disable react/jsx-props-no-spreading */
import { render, screen } from '@testing-library/react';
import React from 'react';
import { categorisedProducts } from '../../utils/mockData';
import Checkout from './Checkout';

describe('Checkout Component', () => {
  afterEach(() => jest.clearAllMocks);
  const mockProps = {
    products: categorisedProducts,
    cartItemsCount: 20,
    updateAllOrders: jest.fn(),
  };
  it('should match the snapshot', () => {
    const { container } = render(<Checkout {...mockProps} />);
    expect(container).toMatchSnapshot();
  });
  it('should render form component', () => {
    render(<Checkout {...mockProps} />);
    const formElement = screen.getByTestId('form');
    expect(formElement.tagName).toBe('FORM');
  });
});
