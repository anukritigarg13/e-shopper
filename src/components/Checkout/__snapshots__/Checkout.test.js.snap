// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout Component should match the snapshot 1`] = `
<div>
  <div
    class="checkout"
  >
    <div
      class="form"
    >
      <form
        data-testid="form"
      >
        Name
        <input
          type="text"
          value=""
        />
        Address
        <input
          type="text"
          value=""
        />
        Contact Number
        <input
          type="number"
          value=""
        />
        E-mail
        <input
          type="text"
          value=""
        />
      </form>
    </div>
    <div>
      <button
        type="button"
      >
        Submit
      </button>
    </div>
    <div
      class="checkout-message"
    >
      <h2 />
    </div>
    <div />
  </div>
</div>
`;
